services/dashboard/overview.go

```diff
@@ -0,0 +1,453 @@
+		if err != nil {
+			web.HTTPError(w, err)
+			return
+		}
+		var resBody web.APIResponse[[]devices.Device]
+		if err := json.NewDecoder(res.Body).Decode(&resBody); err != nil {
+			web.HTTPError(w, err)
+			return
+		}
+		page.Devices = resBody.Data
+
+		if isHX(r) {
+			page.WriteBody(w)
+			return
+		}
+		views.WriteIndex(w, page)
+	}
+}
+
+func deviceDetailPage() http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		device := r.Context().Value("device").(*devices.Device)
+		page := &views.DeviceDetailPage{
+			Device: *device,
+		}
+		if isHX(r) {
+			page.WriteBody(w)
+			return
+		}
+		views.WriteIndex(w, page)
+	}
+}
+
+func sensorDetailPage() http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		device := r.Context().Value("device").(*devices.Device)
+		sensor := r.Context().Value("sensor").(*devices.Sensor)
+
+		res, err := http.Get(fmt.Sprintf("http://core:3000/datastreams?sensor=%d", sensor.ID))
+		if err != nil {
+			web.HTTPError(w, err)
+			return
+		}
+		var resBody web.APIResponse[[]measurements.Datastream]
+		if err := json.NewDecoder(res.Body).Decode(&resBody); err != nil {
+			web.HTTPError(w, err)
+			return
+		}
+
+		page := &views.SensorDetailPage{
+			Device:      *device,
+			Sensor:      *sensor,
+			Datastreams: resBody.Data,
+		}
+		if isHX(r) {
+			page.WriteBody(w)
+			return
+		}
+		views.WriteIndex(w, page)
+	}
+}
+
+func devicesStreamMap() http.HandlerFunc {
+	upgrader := websocket.Upgrader{
+		ReadBufferSize:  1024,
+		WriteBufferSize: 1024,
+	}
+
+	getDevicePage := func(sensorGroupID, cursor string) ([]devices.Device, string, error) {
+		q := url.Values{}
+		q.Set("cursor", cursor)
+		if sensorGroupID != "" {
+			q.Set("sensorgroup", sensorGroupID)
+		}
+		res, err := http.Get("http://core:3000/devices?" + q.Encode())
+		if err != nil {
+			return nil, "", err
+		}
+		var resBody pagination.APIResponse[devices.Device]
+		if err := json.NewDecoder(res.Body).Decode(&resBody); err != nil {
+			return nil, "", err
+		}
+		return resBody.Data, resBody.Links.Next, nil
+	}
+	type Marker struct {
+		DeviceID  int64   `json:"device_id"`
+		Label     string  `json:"label"`
+		Latitude  float64 `json:"latitude"`
+		Longitude float64 `json:"longitude"`
+	}
+
+	return func(w http.ResponseWriter, r *http.Request) {
+		sensorGroupID := r.URL.Query().Get("sensorgroup")
+		ws, err := upgrader.Upgrade(w, r, nil)
+		if err != nil {
+			web.HTTPError(w, err)
+			return
+		}
+
+		go func() {
+			var nextCursor string
+			for {
+				// Start fetching pages of devices and stream them to the client
+				devices, cursor, err := getDevicePage(sensorGroupID, nextCursor)
+				if err != nil {
+					ws.Close()
+					return
+				}
+
+				for _, dev := range devices {
+					if dev.Latitude == nil || dev.Longitude == nil {
+						continue
+					}
+					writer, err := ws.NextWriter(websocket.TextMessage)
+					if err != nil {
+						log.Printf("cannot open writer for ws: %v\n", err)
```

Comment:
this will already generate a lot of logs when 1 client is not functioning correctly

Reply:
Fixed, will now return and close the ws